dev/com.ibm.ws.transport.http/src/com/ibm/ws/http/channel/internal/HttpServiceContextImpl.java

```diff
@@ -3363,78 +3378,42 @@ final protected void sendFullOutgoing(WsByteBuffer[] wsbb) throws IOException {
             }
         }
 
-        DefaultHttpContent content;
         boolean shouldSkipWriteOnUpgrade = nettyResponse.status().equals(HttpResponseStatus.SWITCHING_PROTOCOLS)
                                            && !nettyContext.channel().attr(NettyHttpConstants.PROTOCOL).get().equals("HTTP2");
-        if (!shouldSkipWriteOnUpgrade && Objects.nonNull(buffers) && this.nettyContext.channel().pipeline().get(NettyServletUpgradeHandler.class) == null) {
-            this.addBytesWritten(GenericUtils.sizeOf(buffers));
-            // TODO check this as I believe it is not longer required
-            this.nettyContext.channel().attr(NettyHttpConstants.RESPONSE_BYTES_WRITTEN).set(numBytesWritten);
+        // On upgrade but haven't written headers
+        if(shouldSkipWriteOnUpgrade && sendHeaders) {
+            sendNettyHeaders();
+        }
+        else if (!shouldSkipWriteOnUpgrade && Objects.nonNull(buffers) && this.nettyContext.channel().pipeline().get(NettyServletUpgradeHandler.class) == null) {
+
+            addBytesWritten(GenericUtils.sizeOf(buffers));
 
             if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
                 Tr.debug(tc, "Number of bytes to write: " + getNumBytesWritten());
             }
-            for (WsByteBuffer buffer : buffers) {
-                if (Objects.nonNull(buffer)) {
-                    if (buffer.remaining() == 0) {
-                        if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
-                            Tr.debug(tc, "Ignoring buffer with no content");
-                        }
-                        continue;
-                    }
-                    if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
-                        Tr.debug(tc, "Writing buffer from sendFullOutgoing: " + buffer);
-                    }
-                    String streamId = nettyResponse.headers().get(HttpConversionUtil.ExtensionHeaderNames.STREAM_ID.text(), "-1");
-                    this.nettyContext.channel().write(new StreamSpecificHttpContent(Integer.valueOf(streamId), Unpooled.wrappedBuffer(WsByteBufferUtils.asByteArray(buffer))));
-                }
+
+            String streamId = nettyResponse.headers().get(HttpConversionUtil.ExtensionHeaderNames.STREAM_ID.text(), "-1");
```

Comment:
Favor NettyResponseMessage getHeader

Reply:
Created https://github.com/OpenLiberty/open-liberty/issues/32231 to follow up